### PR TITLE
Implement RFC3137 trim-paths sysroot changes - take 2

### LINKED ISSUE #85463: remap libstd paths in backtraces for test crates
I just got this backtrace when running `RUST_BACKTRACE=1 cargo test` on salsa-rs/salsa@bb4b42acf159332ddf9da735a85cf8d082de882a:

```
thread 'frozen::in_par_get_set_cancelation' panicked at 'called `Result::unwrap()` on an `Err` value: Any { .. }', tests/parallel/frozen.rs:64:33
stack backtrace:
   0: rust_begin_unwind
             at /rustc/881c1ac408d93bb7adaa3a51dabab9266e82eee8/library/std/src/panicking.rs:493:5
   1: core::panicking::panic_fmt
             at /rustc/881c1ac408d93bb7adaa3a51dabab9266e82eee8/library/core/src/panicking.rs:92:14
   2: core::result::unwrap_failed
             at /rustc/881c1ac408d93bb7adaa3a51dabab9266e82eee8/library/core/src/result.rs:1355:5
   3: core::result::Result<T,E>::unwrap
             at /home/jonas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/result.rs:1037:23
   4: parallel::frozen::in_par_get_set_cancelation
             at ./tests/parallel/frozen.rs:64:18
   5: parallel::frozen::in_par_get_set_cancelation::{{closure}}
             at ./tests/parallel/frozen.rs:13:1
   6: core::ops::function::FnOnce::call_once
             at /home/jonas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:227:5
   7: core::ops::function::FnOnce::call_once
             at /rustc/881c1ac408d93bb7adaa3a51dabab9266e82eee8/library/core/src/ops/function.rs:227:5
```

The remapping from `/rustc/...` paths to the rust-src component only seems to kick in *once* here, but I expected it to always work (since the component is installed). Frames 6 and 7 also point to the exact same location, but path remapping still failed for one of them.

### LINKED ISSUE #105907: /rustc/$hash prefix is not being mapped when expected
Spawned off of #64507

I tried this code:

```shell
% rustup default
stable-aarch64-apple-darwin (default)
% rustup component list | grep installed
cargo-aarch64-apple-darwin (installed)
clippy-aarch64-apple-darwin (installed)
rust-src (installed)
rust-std-aarch64-apple-darwin (installed)
rust-std-wasm32-unknown-unknown (installed)
rustc-aarch64-apple-darwin (installed)
rustfmt-aarch64-apple-darwin (installed)
% cat map-panic.rs
fn main() {
    let _x: Option<String> = Some(42u32).map(|_| panic!("hello world"));
}
% rustc -vV
rustc 1.64.0 (a55dd71d5 2022-09-19)
binary: rustc
commit-hash: a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52
commit-date: 2022-09-19
host: aarch64-apple-darwin
release: 1.64.0
LLVM version: 14.0.6
% rustc -g map-panic.rs
% RUST_BACKTRACE=1 ./map-panic
thread 'main' panicked at 'hello world', map-panic.rs:2:50
stack backtrace:
   0: std::panicking::begin_panic
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/panicking.rs:616:12
   1: map_panic::main::{{closure}}
             at ./map-panic.rs:2:50
   2: core::option::Option<T>::map
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/option.rs:929:29
   3: map_panic::main
             at ./map-panic.rs:2:30
   4: core::ops::function::FnOnce::call_once
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/ops/function.rs:248:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

I expected to see this happen: I expected the `/rustc/$hash` to be mapped to the local paths in the installed `rust-src` component, rather than being left as raw `/rustc/$hash` 

Instead, this happened: `/rustc/$hash` is showing up in the output.

### Meta

`rustc --version --verbose`:
```
rustc 1.64.0 (a55dd71d5 2022-09-19)
binary: rustc
commit-hash: a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52
commit-date: 2022-09-19
host: aarch64-apple-darwin
release: 1.64.0
LLVM version: 14.0.6
```

### PR DESCRIPTION
**Pull request by Urgau**
This PR is a continuation of https://github.com/rust-lang/rust/pull/118149. Nothing really changed, except for https://github.com/rust-lang/rust/pull/129408 which I was able to trigger locally.

Original description:

> Implement parts of #111540
> 
> Right now, backtraces into sysroot always shows /rustc/$hash in diagnostics, e.g.
> 
> ```
> thread 'main' panicked at 'hello world', map-panic.rs:2:50
> stack backtrace:
>    0: std::panicking::begin_panic
>              at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/panicking.rs:616:12
>    1: map_panic::main::{{closure}}
>              at ./map-panic.rs:2:50
>    2: core::option::Option<T>::map
>              at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/option.rs:929:29
>    3: map_panic::main
>              at ./map-panic.rs:2:30
>    4: core::ops::function::FnOnce::call_once
>              at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/ops/function.rs:248:5
> note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
> ```
> 
> [RFC 3127 said](https://rust-lang.github.io/rfcs/3127-trim-paths.html#changing-handling-of-sysroot-path-in-rustc)
> 
> > We want to change this behaviour such that, when rust-src source files can be discovered, the virtual path is discarded and therefore the local path will be embedded, unless there is a --remap-path-prefix that causes this local path to be remapped in the usual way.
> 
> This PR implements this behaviour. When `rust-src` is present at compile time, rustc replaces /rustc/$hash with a real path into local rust-src with best effort. To sanitise this, users must explicitly supply `--remap-path-prefix=<path to rust-src>=foo`.

cc @cbeuw
Fix #105907 
Fix #85463

try-job: dist-x86_64-linux
try-job: x86_64-msvc
try-job: dist-x86_64-msvc
try-job: armhf-gnu

### CODE REVIEW

**Comment by Urgau** [comment]
Great, the extended try is successful! `dist-x86_64-linux` and `x86_64-msvc` were the builder that failed in the previous PR: https://github.com/rust-lang/rust/pull/118149#issuecomment-1858901174.

I've verified locally (with [`rustup-toolchain-install-master`](https://github.com/kennytm/rustup-toolchain-install-master)):
 1. that the prebuild `rlib`s only contains `/rustc/` paths ✔
 2. that compiling `tests/codegen/remap_path_prefix/issue-73167-remap-std.rs` *without* the `rust-src` component only leads to `/rustc/` paths ✔
 3. that compiling `tests/codegen/remap_path_prefix/issue-73167-remap-std.rs`  **with** the `rust-src` component leads to many (not all) paths redirected to `/home/gh-Urgau/.rustup/toolchains/a600fe8b275a62c2973e9fc1b589e50302eb2540/lib/rustlib/src/rust/library/` ✔

Everything looks go.
r? compiler

File: src/tools/compiletest/src/header.rs [hunk 1]
```diff
@@ -1115,6 +1115,7 @@ fn expand_variables(mut value: String, config: &Config) -> String {
     const CWD: &str = "{{cwd}}";
     const SRC_BASE: &str = "{{src-base}}";
     const BUILD_BASE: &str = "{{build-base}}";
+    const RUST_SRC_BASE: &str = "{{rust-src-base}}";
```
Comment (jieyouxu) [inline src/tools/compiletest/src/header.rs:1118]: Suggestion: we definitely want to update rustc-dev-guide's compiletest directive remarks to mention `{{rust-src-base}}`, which is not to be confused with `{{src-base}}`.

**Review by jieyouxu — APPROVED** [review]
Thanks, the changes themselves look good to me. We definitely should update rustc-dev-guide to mention existence and purpose of `{{rust-src-base}}`, would be very cool if you can open a follow-up dev-guide PR.

**Comment by jieyouxu** [comment]
Not going to block this PR on the dev-guide `{{rust-src-base}}` docs, but lmw if you intend to document that or if you want me to write it.

File: src/tools/compiletest/src/header.rs (hunk 1 above)
Comment (Urgau) [reply src/tools/compiletest/src/header.rs:1118]: Opened https://github.com/rust-lang/rustc-dev-guide/pull/2079

**Comment by tgross35** [comment]
It looks like the rollup failed at a point that was touched by this PR https://github.com/rust-lang/rust/pull/130893#issuecomment-2377859646

```
2024-09-26T20:19:40.5381850Z ---- [incremental] tests\incremental\unchecked_dirty_clean.rs stdout ----
2024-09-26T20:19:40.5382788Z thread '[incremental] tests\incremental\unchecked_dirty_clean.rs' panicked at src\tools\compiletest\src\runtest.rs:2319:14:
2024-09-26T20:19:40.5383996Z lib/rustlib/src/rust in target is a symlink to checkout root: Os { code: 4390, kind: Uncategorized, message: "The file or directory is not a reparse point." }
```

Seems like that could be the spurious MSVC issue. I'll start a try job, if that passes then we can just try again.

@bors r-

**Comment by tgross35** [comment]
Huh, guess we have a new flavor of https://github.com/rust-lang/rust/issues/127883. Or something else caused the failure, but nothing else from the rollup makes sense.

@bors r=jieyouxu

**Comment by tgross35** [comment]
Same error in a different rollup at https://github.com/rust-lang/rust/pull/130913#issuecomment-2378323801, seems like this must be influencing it somehow.

@bors rollup=iffy

**Comment by jieyouxu** [comment]
The failure looks genuine, I suspect it's the compiletest changes having to do with read_link.

**Comment by Urgau** [comment]
I have looked at all the information (logs, code, timing, ...) available and I have theory that explains all the behavior we are seeing with this PR:

 1. The try build done earlier by @tgross35 for `dist-x86_64-msvc` only took 1h38min total, but in the full CI build it failed at 1h54min, a significant difference. Looking a logs revealed that many of tests weren't run, included the ones that are failing in the full CI build 😲.
    The try build passing is therefore irrelevant.
 2. It's always the same tests under the `incremental` test-suite that fails, looking at them more closely revealed that they contain annotations (e.g. ERROR cannot find function thread_rng), which means that the function `normalize_output` (where the panic happens) is called.
    And I think the `normalize_output` fn is not called when there aren't any annotations.
    Explaining why only those tests fails.
 3. As for `[..]/lib/rustlib/src/rust` not being a symlink, it's probably due to `opt` which unpacks and copies the components instead of symlinking them.
    ```
    Copying directory C:\a\rust\rust\build\unpacked-dist\rust-std-nightly-x86_64-pc-windows-msvc\rust-std-x86_64-pc-windows-msvc\lib\rustlib\x86_64-pc-windows-msvc\lib to C:\a\rust\rust\build\unpacked-dist\rustc-nightly-x86_64-pc-windows-msvc\rustc\lib\rustlib\x86_64-pc-windows-msvc\lib
    ```
 4. We call `read_link` to match what `rustc` is doing otherwise the paths won't be the same. https://github.com/rust-lang/rust/blob/2f8d81f9dbac6b8df982199f69da04a4c8357227/compiler/rustc_session/src/config.rs#L2838-L2850

With all of that, I think it's safe to say that it's not an `MSVC` issue and that the fix should be to not fail if we can't un-symlink, like what `rustc` does.

**Review by jieyouxu — APPROVED** [review]
Thanks

**Comment by jieyouxu** [comment]
I'm not exactly sure *why* opt-dist sysroot path prefixes becomes wonky...

I suspect we may need to modify opt-dist to pass some information to bootstrap -> compiletest to possibly repair the path prefix up to something like unpacked-dist/rustc-nightly-x86_64-unknown-linux-gnu/rustc/lib/rustlib/src/rust, but I haven't dug deeply into this yet.

We wouldn't want to skip these tests naively especially if it can possibly reveal bugs in PGO+LTO among other things.

**Comment by jieyouxu** [comment]
If the opt-dist tests steps never get reached in PR CI or try-jobs, let's rollup=never and send it off for full CI. r=me once you drop the CI commit.

@bors rollup=never (expected to fail)

**Comment by oxalica** [comment]
It's unfortunate that in 1.83, we have this behavior change, but have not stabilized `trim-paths`. I'm expecting to have a simple constant flag like `CARGO_TRIM_PATHS=all` before introducing this change though. This introduced some unexpected runtime dependencies for binary artifacts on Nix (https://github.com/oxalica/rust-overlay/issues/199).

In [the release note](https://github.com/rust-lang/rust/releases/tag/1.83.0#user-content-1.83.0-Compatibility-Notes), it's mentioned that:
> replaces /rustc/$hash with a real path into the local rust-src component with best effort

> users must explicitly supply --remap-path-prefix=<path to rust-src>=foo or not have the rust-src component installed.

So in build scripts, we now need to find out the path to rust-src in the same way as rustc, and add a remap flag. I guess checking and using `$(rustc --print=sysroot)/lib/rustlib/src/rust` should be enough? (as the [`{{rust-src-base}}` impl](https://github.com/rust-lang/rust/pull/129687/files#diff-bf53d26b9ca67283fc825e902a64780b463c153d1b1df0c8e782da0300f08956R1149)) Or is there any overriding flags or envvars we also need to check?

**Comment by Urgau** [comment]
> I guess checking and using `$(rustc --print=sysroot)/lib/rustlib/src/rust` should be enough? 

Yes, I think it should be enough. As for potential env variables (which I don't think there are) or additional flags (only `--sysroot`?) they are reflected in the `--print=sysroot`.

Another possibility (which the release notes also alludes to) is to not have the `rust-src` component. I'm a bit surprised that your build system is having at all, I wouldn't expect it.

**Comment by oxalica** [comment]
> Yes, I think it should be enough. As for potential env variables (which I don't think there are) or additional flags (only `--sysroot`?) they are reflected in the `--print=sysroot`.

Thanks for the info.

> Another possibility (which the release notes also alludes to) is to not have the `rust-src` component. I'm a bit surprised that your build system is having at all, I wouldn't expect it.

I didn't expect it too. Apparently there are someone using a non-minimal toolchain for build system. Yeah I'll comment that it is not the typical toolchain for release build.

**Pull request merged**